Add support for PSV EntryFunctionName

coopp opened the pull request:
This change introduces a version 3 of the PSV data that includes support for the name of the entry function as an offset into StringTable data to a null-terminated utf-8 string. 

Additional tests were added to ensure that the new value was properly serialized/deserialized from object data.

Fixes #80175

Commit:
Add support for PSV EntryFunctionName

python3kgae commented on llvm/include/llvm/MC/DXContainerPSVInfo.h:
```diff
@@ -46,7 +46,7 @@ struct PSVSignatureElement {
 struct PSVRuntimeInfo {
   bool IsFinalized = false;
-  dxbc::PSV::v2::RuntimeInfo BaseData;
+  dxbc::PSV::v3::RuntimeInfo BaseData;
```
Shall we add the [PSVStringTable](https://github.com/microsoft/DirectXShaderCompiler/blob/main/include/dxc/DxilContainer/DxilPipelineStateValidation.h#L484) too?
Since the EntryFunctionName in v3::RuntimeInfo is index to the StringTable.

coopp replied:
I'll look more into this.  It would make sense to have the string table too.  I need to understand what is involved in adding this.

llvm-beanz replied:
The string table already exists, but it is generated in the `write` function in the cpp file.

coopp replied:
Good to know!

llvm-beanz commented on llvm/lib/ObjectYAML/DXContainerYAML.cpp:
```diff
@@ -347,6 +353,11 @@ void DXContainerYAML::PSVInfo::mapInfoForVersion(yaml::IO &IO) {
+  if (Version == 2)
+    return;
+
+  IO.mapRequired("EntryFunctionName", Info.EntryFunctionName);
```
In the YAML this should be a string not an integer. The Integer is the offset into the string table, for the textual dump we want to read this out into a string and be able to rebuild the string table.

The `DirectX::PSVRuntimeInfo` object has a method `getStringTable()`, which returns a StringRef that represents the full string table (just a bunch of packed null-terminated strings). The new field is an offset into that table for where the entry function name is stored.

In `dxcontainer2yaml.cpp` we'll want to translate that offset to a string that is stored in the `DXContainerYAML::PSVInfo` object, and we'll need to pass the string into the MC writer to add it into the string table builder.

coopp replied:
Lastet commit updates to use a string in the yaml and writes the value to a string table.  There was a bit of refactoring that had to be done because I needed the offset of the entrypoint to be written into the runtime info before it was written out.  So I am sure more cleanups are necessary, but I believe this is heading in the correct direction.

Commit:
Updated change to use a string in the YML instead of an offset and added writing of the entry point into a proper string table

llvm-beanz commented on llvm/include/llvm/MC/DXContainerPSVInfo.h:
```diff
@@ -65,18 +68,97 @@ struct PSVRuntimeInfo {
+
+    BaseData.EntryFunctionName =
+        static_cast<uint32_t>(DXConStrTabBuilder.getOffset(EntryFunctionName));
+
```
We should move the `finalize` function body into the cpp file. This has become a non-trivial bit of logic, so we shouldn't keep it in the header.

coopp replied:
ok!.  no problem.

llvm-beanz commented on llvm/include/llvm/MC/DXContainerPSVInfo.h:
```diff
@@ -65,18 +68,97 @@ struct PSVRuntimeInfo {
   SmallVector<uint32_t> InputPatchMap;
   SmallVector<uint32_t> PatchOutputMap;
 
+  StringTableBuilder DXConStrTabBuilder;
```
Any fields that shouldn't be set by the external caller we should make private. We should also look at these to see if they need to be members or if they should just move to the `finalize` function instead of being in the `write` function.

coopp replied:
Ahh.. Good point.  I'll go over this again and see what I can do.

llvm-beanz reviewed (commented):
A few comments. Mostly looking good.

coopp commented on llvm/include/llvm/BinaryFormat/DXContainer.h:
```diff
@@ -424,6 +424,19 @@ struct ResourceBindInfo : public v0::ResourceBindInfo {
+
+namespace v3 {
+struct RuntimeInfo : public v2::RuntimeInfo {
+  uint32_t EntryFunctionName;
```
Sure!.  I'll fix that up!

Commit:
Rename EntryFunctionName to EntryNameOffset

This change makes the variable be more descriptive to what it is actually representing.

Commit:
Move finalize() implementation to .cpp file reducing clutter in the header file.

Commit:
clang-format! Fix'n up the unpretty bits.

Commit:
Move members to private that are not accessed by external setters.

llvm-beanz commented on llvm/lib/MC/DXContainerPSVInfo.cpp:
```diff
@@ -162,6 +97,94 @@ void PSVRuntimeInfo::write(raw_ostream &OS, uint32_t Version) const {
                                llvm::endianness::little);
 }
 
+static constexpr size_t npos = StringRef::npos;
```
Did you change these functions? They're moved from above which makes the diff look a lot larger than probably required. Can you put them back where they were originally to reduce the diff size?

coopp replied:
Yes!  I can fix this up!

Commit:
Reverted move of some helper functions

Commit:
Removed copy of helper code (from moved location)

bogner commented on llvm/include/llvm/BinaryFormat/DXContainer.h:
```diff
@@ -424,6 +424,19 @@ struct ResourceBindInfo : public v0::ResourceBindInfo {
+struct RuntimeInfo : public v2::RuntimeInfo {
+  uint32_t EntryNameOffset;
+
+  void swapBytes() { sys::swapByteOrder(EntryNameOffset); }
```
Does this need to call `v2::RuntimeInfo::swapBytes`?
